### PR TITLE
UefiPayloadPkg: FDT support Phase 1

### PR DESCRIPTION
**Pull request by LinusvPelt**
# Description

Per UPL spec the agreement is to use FDT as communication data format between bootloaders and payloads.
https://universalpayload.github.io/spec/
This PR is adding FDT support to UefiPayloadPkg, meantime, also maintaining backward compatibility with old communication data format - HOB.
Several new library are created mainly for different arch support so the UPL main flow can be still common cross different arch.
A PCD "PcdHandOffFdtEnable" is created for switching either FDT or HOB during the build time to reduce binary size impact.
(Currently PcdHandOffFdtEnable=FALSE as default for backward compatibility)

Note: FDT support is added only to FIT header UPL build, the old ELF header UPL build remains the same with only HOB support.

<_For each item, place an "x" in between `[` and `]` if true. Example: `[x]` (you can also check items in GitHub UI)_>

<_Create the PR as a Draft PR if it is only created to run CI checks._>

<_Delete lines in \<\> tags before creating the PR._>

- [x] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

Tested by internal platform for booting Windows.

## Integration Instructions

If platform is ready to switch to FDT model, platform DSC will have to hook new library "BuildFdtLib" to FitPayloadLoaderPeim.inf
Platform also needs to set PcdHandOffFdtEnable to TRUE (can be done by adding --pcd parameter as part of the UniversalPayloadBuild.py command)

### CODE REVIEW

**Comment by mdkinney** [comment]
Please update the title and the description of this PR to provide a summary of the change being made and its impacts.

File: UefiPayloadPkg/UefiPayloadPkg.dsc [hunk 1]
```diff
@@ -471,6 +473,8 @@
   gEfiMdeModulePkgTokenSpaceGuid.PcdVpdBaseAddress|0x0
   gEfiMdeModulePkgTokenSpaceGuid.PcdStatusCodeUseMemory|FALSE
   gEfiMdeModulePkgTokenSpaceGuid.PcdUse1GPageTable|TRUE
+  gUefiPayloadPkgTokenSpaceGuid.PcdHandOffFdtEnable|TRUE
```
Comment (ChaselChiu) [inline UefiPayloadPkg/UefiPayloadPkg.dsc:476]: Set PcdHandOffFdtEnable to FALSE as default for now, platforms ready to use FDT will toggle this PCD as part of the UPL build command (--pcd NAME=VALUE)

File: MdePkg/Library/BaseFdtLib/BaseFdtLib.inf [hunk 2]
```diff
@@ -57,6 +57,9 @@
   BaseMemoryLib
 
 [BuildOptions]
-  MSFT:*_*_IA32_CC_FLAGS = /wd4146 /wd4245
-  MSFT:*_*_X64_CC_FLAGS  = /wd4146 /wd4244 /wd4245 /wd4267
+# warning C4706: assignment within conditional expression
+# if ((err = fdt_splice_(fdt, p, oldlen, newlen)))
+# in BaseFdtLib\libfdt\libfdt\fdt_rw.c (wait for sub module update to remove this)
```
Comment (ChaselChiu) [inline MdePkg/Library/BaseFdtLib/BaseFdtLib.inf:62]: Same change here has been approved by Liming in another PR:
https://github.com/tianocore/edk2/pull/6094#issuecomment-2309292299

File: UefiPayloadPkg/UefiPayloadPkg.dsc (hunk 1 above)
Comment (ChaselChiu) [reply UefiPayloadPkg/UefiPayloadPkg.dsc]: resolved

File: MdePkg/Library/BaseFdtLib/BaseFdtLib.inf (hunk 2 above)
Comment (ChaselChiu) [reply MdePkg/Library/BaseFdtLib/BaseFdtLib.inf:62]: just FYI

File: UefiPayloadPkg/PayloadLoaderPeim/FitPayloadLoaderPeim.inf [hunk 3]
```diff
@@ -35,25 +35,33 @@
...
   PcdLib
-  MemoryAllocationLib
   BaseMemoryLib
   PeiServicesLib
   HobLib
   BaseLib
   PeimEntryPoint
   DebugLib
   FdtLib
+##for testing, remove this for HandOffFdtEnable == FALSE scenario: BuildFdtLib
```
Comment (ChaselChiu) [inline UefiPayloadPkg/PayloadLoaderPeim/FitPayloadLoaderPeim.inf:45]: let use some comments like below example to guide users what to do for this bootloader module:
## Hook this lib to FitPayloadLoaderPeim.inf in platform DSC when HandOffFdtEnable == TRUE.

**Review by ChaselChiu — APPROVED** [review]

File: UefiPayloadPkg/PayloadLoaderPeim/FitPayloadLoaderPeim.inf (hunk 3 above)
Comment (ChaselChiu) [reply UefiPayloadPkg/PayloadLoaderPeim/FitPayloadLoaderPeim.inf:45]: resolved

**Review by gguo11837463 — APPROVED** [review]
Ack-by me.

**Pull request merged**